### PR TITLE
Adds basic infrastructure for NN models

### PR DESCRIPTION
**Pull request by EiffL**
This draft PR aims to solve one of our problems where NNs are defined in multiple places and can be different. 

Here we put the NN defintion in the python module, so that we can import them in notebooks and scripts and they will always correspond to the same thing

@justinezgh can you take a look at this draft PR, check if this code works and add anytihng that it's missing but you know you will need?

### CODE REVIEW

File: sbids/models.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import tensorflow_probability as tfp; tf = tfp.experimental.substrates.jax
+tfb = tfp.bijectors
+tfd = tfp.distributions
+
+from sbids.bijectors import MixtureAffineSigmoidBijector
+
+class AffineSigmoidCoupling(hk.Module):
+  """This is the coupling layer used in the Flow."""
+  def __init__(self, y, *args, layers=[128, 128], n_components=32, activation=jax.nn.silu, **kwargs):
```
Comment (EiffL) [inline sbids/models.py:13]: so why here do we use silu again?

Comment (Justinezgh) [reply sbids/models.py:13]: I don't remember :/

File: sbids/models.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    coupling_layer, list of coupling layers
+    """
+    self.d = d
+    self.n_layer = n_layers
+    self.bijector_fn = bijector_fn
+    super(ConditionalRealNVP, self).__init__(*args, **kwargs)
+    
+  def __call__(self, y):
+    chain = tfb.Chain([
+          tfb.Permute(jnp.arange(self.d)[::-1])(tfb.RealNVP(2, bijector_fn=self.bijector_fn(y, name = 'b%d'%i))) for i in range(self.n_layer)
```
Comment (Justinezgh) [inline sbids/models.py:67]:
```python 
tfb.Permute(jnp.arange(self.d)[::-1])(tfb.RealNVP(self.d//2, bijector_fn=self.bijector_fn(y, name = 'b%d'%i))) for i in range(self.n_layer)
```
 ?

Comment (Justinezgh) [reply sbids/models.py:67]: and can we add the shift and scale bijector ?

Comment (EiffL) [reply sbids/models.py:67]: As part of the Flow, maybe not. This flow is intended to be generic and usable on any problem. The only shift and scale that would make sense is to recenter on 0 and roughly standard deviation. Then on a problem-by-problem we could shift and scale the problem back to a roughly normal distribution before sending it to the flow

Comment (EiffL) [reply sbids/models.py:67]: And that line of code, what it does us to repeat a combination of permutation and realnvp layers as many times as requested

Comment (Justinezgh) [reply sbids/models.py:67]: yes but isn't it self.d//2 instead of 2 in `tfb.RealNVP()` ?

Comment (EiffL) [reply sbids/models.py:67]: Ah yes indeed ^^ sorry, I didn't understand the problem from your question mark ?

Comment (Justinezgh) [reply sbids/models.py:67]: ok except this everything works :)

Comment (EiffL) [inline sbids/models.py:67]:
```suggestion
          tfb.Permute(jnp.arange(self.d)[::-1])(tfb.RealNVP(self.d//2, bijector_fn=self.bijector_fn(y, name = 'b%d'%i))) for i in range(self.n_layer)
```

**Review by Justinezgh — APPROVED** [review]

**Pull request merged**